pkg/channel/message_receiver.go

```diff
@@ -0,0 +1,180 @@
+	}
+	for _, opt := range opts {
+		if err := opt(receiver); err != nil {
+			return nil, err
+		}
+	}
+	return receiver, nil
+}
+
+// Start begins to receive events for the receiver.
+//
+// Only HTTP POST requests to the root path (/) are accepted. If other paths or
+// methods are needed, use the HandleRequest method directly with another HTTP
+// server.
+func (r *MessageReceiver) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- r.httpBindingsReceiver.StartListen(ctx, r)
+	}()
+
+	// Stop either if the receiver stops (sending to errCh) or if the context Done channel is closed.
+	select {
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		break
+	}
+
+	// Done channel has been closed, we need to gracefully shutdown r.ceClient. The cancel() method will start its
+	// shutdown, if it hasn't finished in a reasonable amount of time, just return an error.
+	cancel()
+	select {
+	case err := <-errCh:
+		return err
+	case <-time.After(shutdownTimeout):
+		return errors.New("timeout shutting down http bindings receiver")
+	}
+}
+
+func (r *MessageReceiver) ServeHTTP(response nethttp.ResponseWriter, request *nethttp.Request) {
+	if request.Method != nethttp.MethodPost {
+		response.WriteHeader(nethttp.StatusMethodNotAllowed)
+		return
+	}
+
+	// tctx.URI is actually the path...
+	if request.URL.Path != "/" {
+		response.WriteHeader(nethttp.StatusNotFound)
+		return
+	}
+
+	// The response status codes:
+	//   202 - the event was sent to subscribers
+	//   404 - the request was for an unknown channel
+	//   500 - an error occurred processing the request
+
+	host := request.Host
+	r.logger.Debug("Received request", zap.String("host", host))
+	channel, err := r.hostToChannelFunc(host)
+	if err != nil {
+		r.logger.Info("Could not extract channel", zap.Error(err))
+		response.WriteHeader(nethttp.StatusInternalServerError)
+		return
+	}
+	r.logger.Debug("Request mapped to channel", zap.String("channel", channel.String()))
+
+	message := http.NewMessageFromHttpRequest(request)
+
+	if message.ReadEncoding() == binding.EncodingUnknown {
+		r.logger.Info("Cannot determine the cloudevent message encoding")
+		response.WriteHeader(nethttp.StatusBadRequest)
+		return
+	}
+
+	transformers := append(defaultTransformers, AddHistory(host)...)
+	span := trace.FromContext(request.Context())
+	if span != nil {
+		transformers = append(transformers, tracing.AddTraceparent(span)...)
+	}
+
+	// TODO(slinkydeveloper) Why do i buffer the message if i don't need it (kafka channel use case)?
+	// CopyMessage generates a message that we can reuse several times
+	bufferedMessage, err := buffering.CopyMessage(r.messageBufferingCtx, message, transformers)
+	// We can forget the original message (in case of http, close the body)
+	_ = message.Finish(nil)
+
+	err = r.receiverFunc(request.Context(), channel, bufferedMessage, utils.PassThroughHeaders(request.Header))
```

Comment:
yes this is exactly why we did what we did in the cloudevents sdk, so integrators like Knative can write this kind of code.